lib/Messenger/Handler/AssetUpdateTasksHandler.php

```diff
@@ -51,51 +56,62 @@ public function __invoke(AssetUpdateTasksMessage $message): void
         }
 
         $this->longRunningHelper->deleteTemporaryFiles();
+        $this->lockFactory->createLock($asset->getUpdateQueueLockId())->release();
     }
 
-    private function saveAsset(Asset $asset): void
+    private function saveAsset(Asset $asset, array $saveParams = []): void
     {
         Version::disable();
         $asset->markFieldDirty('modificationDate'); // prevent modificationDate from being changed
-        $asset->save();
+        $asset->save($saveParams);
         Version::enable();
     }
 
     private function processDocument(Asset\Document $asset): void
     {
+        $save = false;
+        $saveParams = [];
+        $asset->removeCustomSetting(Asset::CUSTOM_SETTING_UPDATE_TASK_PROCESSING_FAILED);
         if ($asset->getMimeType() === 'application/pdf' && $asset->checkIfPdfContainsJS()) {
-            $asset->save(['versionNote' => 'PDF scan result']);
+            $save = true;
+            $saveParams['versionNote'] = 'PDF scan result';
         }
 
-        $savedNeeded = false;
         $pageCount = $asset->getCustomSetting('document_page_count');
         if (!$pageCount || $pageCount === 'failed') {
-            $savedNeeded |= $asset->processPageCount();
-        }
+            if (!$asset->processPageCount() || $asset->getCustomSetting('document_page_count') === 'failed') {
+                $asset->setCustomSetting(Asset::CUSTOM_SETTING_UPDATE_TASK_PROCESSING_FAILED, true);
+                $this->logger->warning(sprintf('Failed processing page count for document asset %s.', $asset->getId()));
+
+            } else {
+                $asset->getImageThumbnail(Asset\Image\Thumbnail\Config::getPreviewConfig())->generate(false);
+            }
 
-        if ($savedNeeded) {
-            $this->saveAsset($asset);
+            $save = true;
         }
 
-        if ($asset->getCustomSetting('document_page_count') !== 'failed') {
-            $asset->getImageThumbnail(Asset\Image\Thumbnail\Config::getPreviewConfig())->generate(false);
-        } elseif ($savedNeeded) {
-            $this->logger->warning(sprintf('Failed processing page count for document asset %s.', $asset->getId()));
+        if ($save) {
+            $this->saveAsset($asset, $saveParams);
         }
     }
 
     private function processVideo(Asset\Video $asset): void
     {
+        $asset->removeCustomSetting(Asset::CUSTOM_SETTING_UPDATE_TASK_PROCESSING_FAILED);
+        $failed = true;
+
         if ($duration = $asset->getDurationFromBackend()) {
             $asset->setCustomSetting('duration', $duration);
-        } else {
-            $asset->removeCustomSetting('duration');
+            if ($dimensions = $asset->getDimensionsFromBackend()) {
```

Comment:
initially though that we may not need a nested IF there

```suggestion
            $dimensions = $asset->getDimensionsFromBackend();
            $duration = $asset->getDurationFromBackend();
            
            if ($dimensions && $duration) {
                $asset->setCustomSetting('duration', $duration);
```

Reply:
Since it looks like they are less "independent" from each other, what about having a new method to `getMetaDataFromBackend()` (or so) that returns an associative array `['dimension'=>null, 'duration' => null]`

If i got correctly, by looking at
https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/models/Asset/Video.php#L161-L191

It calls [getDefaultAdapter](https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video.php#L61-L80) (which has class exists check and initializes a new Adapter class
) at least 4 times (when processing is not considered to be failed) that might cause memory issues


more specifically in https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video.php#L52-L56 and getInstance https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video.php#L39

Reply:
> initially though that we may not need a nested IF there

I went for nested if, since it doesn't make sense to read dimensions if reading the duration failed already so save resources.

Reply:
yes, agreeing with the nested here, but wdyt about this https://github.com/pimcore/pimcore/pull/18679#discussion_r2371387229 ? it's needing 4 instantiation of the adapter, not strictly for this PR but might be causing memory issues

Reply:
@kingjia90 yeah, you're right, looking at it, I'd see this as a follow up optimization. 
I think it would also make sense to introduce an interface for adapters and get rid of the [abstract ](https://github.com/pimcore/pimcore/blob/df14e025d3db615fc75b190e3c9f3cc1d68d7304/lib/Video/Adapter.php#L19)instead. 

I'll create a follow up bc-break ticket for that.

Reply:
done #18702